brave-grpc/src/main/java/com/github/kristofa/brave/grpc/BraveGrpcClientInterceptor.java

```diff
@@ -0,0 +1,135 @@
+import com.github.kristofa.brave.SpanId;
+import com.github.kristofa.brave.internal.Nullable;
+import com.google.common.base.Strings;
+import com.twitter.zipkin.gen.Span;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.Status.Code;
+
+import java.util.Collection;
+import java.util.Collections;
+
+public final class BraveGrpcClientInterceptor implements ClientInterceptor {
+
+    private final String clientService;
+    private final ClientRequestInterceptor clientRequestInterceptor;
+    private final ClientResponseInterceptor clientResponseInterceptor;
+    private final ClientSpanThreadBinder clientSpanThreadBinder;
+
+    public BraveGrpcClientInterceptor(String clientService, Brave brave) {
+        if (Strings.isNullOrEmpty(clientService)) {
+            throw new IllegalStateException("clientService name cannot be empty");
+        }
+        this.clientService = clientService;
+        this.clientRequestInterceptor = checkNotNull(brave.clientRequestInterceptor());
+        this.clientResponseInterceptor = checkNotNull(brave.clientResponseInterceptor());
+        this.clientSpanThreadBinder = checkNotNull(brave.clientSpanThreadBinder());
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(final MethodDescriptor<ReqT, RespT> method,
+                                                               final CallOptions callOptions, final Channel next) {
+        return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+
+            @Override
+            public void start(Listener<RespT> responseListener, Metadata headers) {
+                clientRequestInterceptor.handle(new GrpcClientRequestAdapter<>(clientService, method, headers));
+                final Span currentClientSpan = clientSpanThreadBinder.getCurrentClientSpan();
+                super.start(new SimpleForwardingClientCallListener<RespT>(responseListener) {
+                    @Override
+                    public void onClose(Status status, Metadata trailers) {
+                        try {
+                            clientSpanThreadBinder.setCurrentSpan(currentClientSpan);
+                            clientResponseInterceptor.handle(new GrpcClientResponseAdapter(status));
+                            super.onClose(status, trailers);
+                        } finally {
+                            clientSpanThreadBinder.setCurrentSpan(null);
+                        }
+                   }
+                }, headers);
+            }
+        };
+
+    }
+
+    private static final class GrpcClientRequestAdapter<ReqT, RespT> implements ClientRequestAdapter {
```

Comment:
Was there a reason why you did not use the Http abstractions like `HttpClientRequestAdapter` instead of `ClientRequestAdapter`?  The advantage of using it is that the conversion span <-> http is done for you and same across implementations. You will probably also end up with less code.